examples/step-79/step-79.cc

```diff
@@ -0,0 +1,2755 @@
+    // with `using namespace` as above:
+    nonlinear_solution.reinit(block_sizes);
+    system_rhs.reinit(block_sizes);
+
+    {
+      using namespace SolutionBlocks;
+
+      for (unsigned int k = 0; k < n_u; ++k)
+        {
+          nonlinear_solution.block(displacement)[k]            = 0;
+          nonlinear_solution.block(displacement_multiplier)[k] = 0;
+        }
+      for (unsigned int k = 0; k < n_p; ++k)
+        {
+          nonlinear_solution.block(density)[k]            = density_ratio;
+          nonlinear_solution.block(unfiltered_density)[k] = density_ratio;
+          nonlinear_solution.block(unfiltered_density_multiplier)[k] =
+            density_ratio;
+          nonlinear_solution.block(density_lower_slack)[k] = density_ratio;
+          nonlinear_solution.block(density_lower_slack_multiplier)[k] = 50;
+          nonlinear_solution.block(density_upper_slack)[k] = 1 - density_ratio;
+          nonlinear_solution.block(density_upper_slack_multiplier)[k] = 50;
+        }
+    }
+  }
+
+
+  // @sect3{Creating the filter matrix}
+
+  // Next up, a function that is used once at the beginning of the
+  // program: It creates a matrix $H$ so that the filtered density
+  // vector equals $H$ times the unfiltered density.  The creation
+  // of this matrix is non-trivial, and it is used in every
+  // iteration, and so rather than reforming it as we do with the
+  // Newton matrix, it is made only once and stored separately.
+  //
+  // The way this matrix is computed follows the outline used above
+  // already to form its sparsity pattern. We repeat this process here
+  // for the sparsity pattern of this separately formed matrix, and
+  // then actually build the matrix itself. You may want to check the
+  // definition of this matrix in the introduction to this program.
+  template <int dim>
+  void SANDTopOpt<dim>::setup_filter_matrix()
+  {
+    // The sparsity pattern of the filter has already been determined
+    // and implemented in the setup_system() function. We copy the
+    // structure from the appropriate block and use it again here.
+
+    filter_sparsity_pattern.copy_from(
+      sparsity_pattern.block(SolutionBlocks::unfiltered_density,
+                             SolutionBlocks::unfiltered_density_multiplier));
+    filter_matrix.reinit(filter_sparsity_pattern);
+
+    // Having so built the sparsity pattern, now we re-do all of
+    // these loops to actually compute the necessary values of the
+    // matrix entries:
+    for (const auto &cell : dof_handler.active_cell_iterators())
+      {
+        const unsigned int i = cell->active_cell_index();
+
+        std::set<unsigned int>                               neighbor_ids;
+        std::set<typename Triangulation<dim>::cell_iterator> cells_to_check;
+
+        neighbor_ids.insert(i);
+        cells_to_check.insert(cell);
+
+        unsigned int n_neighbors = 1;
+        filter_matrix.add(i, i, filter_r);
+        while (true)
```

Comment:
wouldn't it be easier and faster to iterate over the sparsity pattern instead of figuring out the close cells again?

Reply:
I would need a way to get access to the neighboring cell (and specifically its center point) from the neighboring cell's index to do this, which I don't know how to do. This is because I then use the distance between the two cells. Is there a nice way to do that?